### DISCUSSION TITLE
Features I'm missing from Sway

### DISCUSSION THREAD

**Discussion by hernan3009** [qtile/qtile#2423] [General]
Hi,

I am trying to implement some functions in my config.py file to add some functionality that I miss from Sway and I really wish to have in Qtile. The are:

1) Swap between tiled windows using mouse: Hold MOD+Left clic and drag win_1 to win_2, after releasing win_1 and win_2 must permute their places.
2) Resize tiled windows using mouse: MOD+Right clic in any part of a tiled window lets change its size (and adjust the size of the remaining windows in the group) [1 and 2 can be viewed in action in [this](https://swaywm.org/) video]
3) Stick a window to some region of the monitor:  The window remains always visible in the same way no mater if other windows are modified or if I move to another group (this is nice to play videos will doing other activities...).

I have been many hours reading the Docs and playing with the config file but I am stuck. I am trying to follow the example from [here](https://github.com/qtile/qtile-examples/blob/master/bglasber.py).

I made a keybinding to a new function:
`    Key([mod], "i", lazy.function(new_function), desc="Prueba"),`
The function is:
```
def new_function(qtile):
    if qtile.currentWindow is not None:
        i = qtile.groups.index(qtile.currentGroup)
        qtile.currentWindow.togroup(qtile.groups[i + 1].name)
    return
```
I get (in the .log file)
`AttributeError: 'Qtile' object has no attribute 'currentWindow'`

I managed to send some data to `new_function` through optional parameters like `Group` but I cannot get info and processing it in a reasonable way (as the function in the example). Is this behaviour normal or is some bug?

Qtile version:    0.17.1

**Comment by tych0** [comment]
A while ago we changed a bunch of camelCase stuff to snake_case, this is one of them. You should be able to use current_window.

**Comment by hernan3009** [comment]
Thank you! I continued trying and found a partial solution: `currentWindow` seems to be `current_window` (maybe outdated example file?). The same applies when changing "Window" to "Group".

Are there any known solutions to the three mentioned functions? (I am not a developer and it seems to me that it will be very hard for me to implement them)

**Comment by ramnes** [comment]
1 is already implemented but non-functional at the moment. See PR #1334 for relevant history, discussion, and a fix proposal.

I don't think you will be able to implement 2 globally with our current layouts architecture. That being said, if you're coming from Sway and want something that feels similar, you might be interested in the BSP layout. Modifying this layout specifically to add the feature you mentioned in 2 shouldn't be too hard.

For 3, you just have to bind [`lazy.window.static()`](http://docs.qtile.org/en/latest/manual/ref/commands.html#libqtile.backend.x11.window.Window.cmd_static).

**Reply by tusqasi** [reply]
Why isn't `lay.window.static()` listed in lazy functions list [here](http://docs.qtile.org/en/latest/manual/config/lazy.html) in the docs?

**Reply by m-col** [reply]
You want to look here: http://docs.qtile.org/en/latest/manual/ref/commands.html#window

That other page is just show some examples, though this does look misleading, and it ought to have a link to the page with them all.

**Comment by strangr** [comment]
@ramnes I tried making it sticky with static but the window becomes unresponsive
`Key([mod, "control"], "s", self.test_func(0)),`
```
  def test_func(self, screen):
      @lazy.function
      def f(qtile):
          window = qtile.current_window
          window.cmd_static(
              0,
              0,
              0,
              500,
              400)

      return f
```

Qtile version 0.17.0-1

**Reply by hernan3009** [reply]
@strangr I had similar issues when using `static` (I am the one who opened the original issue). Please, see my the 'NOTE'  in [this issue](https://github.com/qtile/qtile/issues/2424) for a work around solution (truly, I do not know how what is the supposed/desirable/intended behaviour of a `static` window, so I did not open an issue about it).

**Reply by strangr** [reply]
@hernan3009 I saw your post and definitely going to try. just when I saw this suggestion about using static (and I was already experimenting with it), decided to make a comment

**Reply by strangr** [reply]
@hook.subscribe.setgroup
    def setgroup():
        windows = qtile.windows_map.values()
        for window in windows:
            if 'sticky' not in window.cmd_hints():
                continue

            screen_id = window.hints["sticky"]
            if qtile.current_screen.index != screen_id:
                continue

            cw = qtile.groups.index(qtile.current_group)
            window.togroup(qtile.groups[cw].name)

    @hook.subscribe.focus_change
    def focus_change():
        windows = qtile.windows_map.values()
        for window in windows:
            if 'sticky' in window.cmd_hints():
                window.cmd_bring_to_front()


    def toggle_sticky(self):
        @lazy.function
        def f(qtile):
            if 'sticky' in qtile.current_window.cmd_hints():
                qtile.current_window.cmd_disable_floating()
                qtile.current_window.hints.pop("sticky", None)
            else:
                qtile.current_window.cmd_bring_to_front()
                qtile.current_window.cmd_enable_floating()
                qtile.current_window.hints["sticky"] = qtile.current_screen.index

        return f

@hernan3009 this is how I did it, I have multiple screens and I wanted for sticky window to stay on the same screen.
have not done changing screen yet.

**Reply by hernan3009** [reply]
@strangr Nice! today I added a second monitor to my PC so your script will be very useful to me!
I really like your implementation, although I do not know what "@lazy.function" does. I guess that this has some effects on the way in which the binding is done, mine is:
`    Key([mod], "i", lazy.function(stay_there), desc="Prueba"),`
What is yours?
Thank you very much

**Reply by strangr** [reply]
@hernan3009 yes, this was the only way I managed to make it working while having access to qtile. also not sure why :)
here is my implementation
https://github.com/strangr/dotfiles/blob/ebce8851f003a42ba51dafadd177f32e9d2041d2/.config/qtile/QScreen.py

    Key([mod, "control"], "s", self.toggle_sticky()),

**Reply by tusqasi** [reply]
This is supposed to stick the window to the screen, right?
It is not working, am i missing anything? I copied the whole thing, and added imports. It still didn't work. I tried both solutions from @strangr and @hernan3009.
I also tried `window.static()` which made windows unresponsive.
